### PR TITLE
feat: throw error when bucket name contains "/"

### PR DESCRIPTION
**Pull request by AllanFly120**
*Issue #, if available:*
https://github.com/aws/aws-sdk-js/issues/1046
https://github.com/aws/aws-sdk-js/issues/1035
https://github.com/aws/aws-sdk-js/issues/1572

*Description of changes:*
S3 client will throw error if the `Bucket` parameter contains `/` that would interfere the signature V4.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/client-s3-node/package.json [hunk 1]
```diff
@@ -57,12 +58,12 @@
     "tslib": "^1.8.0"
   },
   "devDependencies": {
-    "@aws-sdk/client-documentation-generator": "^0.1.0-preview.2",
+    "@aws-sdk/client-documentation-generator": "^0.1.0-preview.1",
     "@types/jest": "^24.0.12",
-    "@types/node": "^10.0.0",
```
Comment (trivikr) [inline packages/client-s3-node/package.json]: Looks like your workspace isn't synced from upstream as this was fixed in https://github.com/aws/aws-sdk-js-v3/pull/301
Can you sync your workspace and regenerate S3 client?

File: packages/middleware-sdk-s3/src/validate-bucket-name.spec.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  const mockNextHandler = jest.fn();
+  const composedHandler: SerializeHandler<any, any> = validateBucketName(
+    mockNextHandler
+  );
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it("throws error if Bucket parameter contains '/'", async () => {
```
Comment (trivikr) [inline packages/middleware-sdk-s3/src/validate-bucket-name.spec.ts]: Add a unit test case where Bucket doesn't contain '/' and no error is thrown

File: packages/middleware-sdk-s3/src/validate-bucket-name.spec.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        }
+      });
+    } catch (e) {
+      error = e;
+    }
+    expect(error).toBeDefined();
+    expect(error.message).toEqual(
+      `Bucket name shouldn't contain '/', received '${bucket}'`
+    );
+    expect(error.name).toEqual("InvalidBucketName");
```
Comment (trivikr) [inline packages/middleware-sdk-s3/src/validate-bucket-name.spec.ts]:
```suggestion
    expect(error.name).toEqual("InvalidBucketName");
    expect(mockNextHandler.mock.calls.length).toBe(1);
```

File: packages/middleware-sdk-s3/src/validate-bucket-name.spec.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it("throws error if Bucket parameter contains '/'", async () => {
+    const bucket = "bucket/part/of/key";
+    let error;
+    try {
+      await composedHandler({
```
Comment (trivikr) [inline packages/middleware-sdk-s3/src/validate-bucket-name.spec.ts]: The following can be stored in a parameter say `args`
```javascript
const args = {
  input: {
    Bucket: bucket
  }
}
```

This way you can add following check at the end of the file:
```javascript
expect(mockNextHandler.mock.calls[0][0]).toBe(args);
```

File: packages/client-s3-node/package.json (hunk 1 above)
Comment (AllanFly120) [reply packages/client-s3-node/package.json]: Nice catch

**Pull request merged**